BinaryDataFileRandomAccess.truncate() leaves writePosition stale, producing zero-gap corruption in nodes-data.obj

### Version

5.6.0 

### What happened?

- Module: `jena-dboe-base` (`org.apache.jena.dboe.base.file.BinaryDataFileRandomAccess`).
`BinaryDataFileRandomAccess.truncate(length)` in `jena-dboe-base` calls `RandomAccessFile.setLength(length)` but does not reset the class's own `writePosition` field. A later `switchToWriteMode()` (triggered by any read→write transition, e.g. a write after a read inside the same `BinaryDataFileRandomAccess` lifecycle) seeks the underlying `RandomAccessFile` to the stale `writePosition` past the just-truncated EOF. The next `file.write(...)` extends the file at that position; the filesystem zero-fills the range between the truncated length and the stale seek, leaving a contiguous zero gap inside the file.

What I expected: after `truncate(length)` followed by `read(...)` then `write(buf)`, the write should land at offset `length`, and the file should contain the written bytes only (no zero padding).

What actually happens: the write lands at the stale `writePosition` (the file length *before* `truncate`), and the filesystem zero-fills the intervening bytes.

Minimal reproducer, no framework:

    var file = new BinaryDataFileRandomAccess("/tmp/bdf-bug.bin");
    file.open();
    byte[] ab = new byte[1024]; java.util.Arrays.fill(ab, (byte) 0xAB);
    for (int i = 0; i < 300; i++) file.write(ab, 0, 1024); // 300 KB of 0xAB
    file.truncate(100 * 1024);                             // file truncated to 100 KB
    file.read(0, new byte[16], 0, 16);                     // flips internal readMode
    byte[] cd = new byte[100]; java.util.Arrays.fill(cd, (byte) 0xCD);
    long pos = file.write(cd, 0, 100);
    file.sync();
    file.close();
    // Expect: pos == 102400, file is 102500 bytes, no zero bytes.
    // Actual: pos == 307200, file is 307300 bytes, 204800 bytes of 0x00 between
    //         the 100 KB of 0xAB and the 100 bytes of 0xCD.

Under TDB2 this surfaces as zero regions inside `nodes-data.obj` after an aborted write transaction (`TransBinaryDataFile._abort()` calls `truncate()`) followed by a transaction whose first access to that file is a read (typical for SPARQL UPDATE — node lookups before allocation). `NodeTable` reads that later resolve a NodeId into a zero region throw `TDBException: NodeTableTRDF/Read → TProtocolException: Unrecognized type 0`. The store is persistently corrupt until the affected range is overwritten; a Fuseki restart does not recover it.

I have a one-line fix and a JUnit 5 regression test ready as a PR against `jena-db/jena-dboe-base` — will open it once this issue has a number to reference.

### Relevant output and stacktrace

```shell
# Output of the minimal reproducer above, on Jena 5.6.0:

after 300 KB writes:   writePosition=307200, file on disk=307200 bytes
after truncate(100K):  writePosition=307200 (stale), file on disk=102400 bytes
after read of 16 B:    readMode=true
after write of 100 B:  returned pos=307200 (should be 102400), file on disk=307300 bytes
byte-content audit:    102400 × 0xAB, 204800 × 0x00, 100 × 0xCD


# The TDB2-level exception this produces when a later NodeTable read resolves a
# NodeId whose offset falls inside one of the zero regions in nodes-data.obj:

org.apache.jena.tdb2.TDBException: NodeTableTRDF/Read
    at org.apache.jena.tdb2.store.nodetable.NodeTableTRDF.readNodeFromTable(NodeTableTRDF.java:87)
    at org.apache.jena.tdb2.store.nodetable.NodeTableNative._retrieveNodeByNodeId(NodeTableNative.java:102)
    at org.apache.jena.tdb2.store.nodetable.NodeTableNative.getNodeForNodeId(NodeTableNative.java:52)
    at org.apache.jena.tdb2.store.nodetable.NodeTableCache._retrieveNodeByNodeId(NodeTableCache.java:212)
    at org.apache.jena.tdb2.store.nodetable.NodeTableCache.getNodeForNodeId(NodeTableCache.java:137)
    at org.apache.jena.tdb2.store.nodetable.NodeTableWrapper.getNodeForNodeId(NodeTableWrapper.java:52)
    ...
Caused by: org.apache.thrift.protocol.TProtocolException: Unrecognized type 0
    at org.apache.thrift.protocol.TCompactProtocol.getTType(...)
    ...
```

### Are you interested in making a pull request?

Yes

--- a/jena-db/jena-dboe-base/src/main/java/org/apache/jena/dboe/base/file/BinaryDataFileRandomAccess.java
+++ b/jena-db/jena-dboe-base/src/main/java/org/apache/jena/dboe/base/file/BinaryDataFileRandomAccess.java
@@ -103,7 +103,10 @@ public class BinaryDataFileRandomAccess implements BinaryDataFile {
     public void truncate(long length) {
         checkOpen();
         switchToWriteMode();
-        try { file.setLength(length); }
+        try {
+            file.setLength(length);
+            writePosition = length;
+        }
         catch (IOException ex) { IO.exception(ex); }
     }
 

--- a/jena-db/jena-dboe-base/src/test/java/org/apache/jena/dboe/base/file/TestBinaryDataFileRandomAccessTruncate.java
+++ b/jena-db/jena-dboe-base/src/test/java/org/apache/jena/dboe/base/file/TestBinaryDataFileRandomAccessTruncate.java
@@ -1,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ *   SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.apache.jena.dboe.base.file;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.Arrays;
+
+import org.apache.jena.atlas.lib.FileOps;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Regression test: {@link BinaryDataFileRandomAccess#truncate(long)} must reset
+ * the internal {@code writePosition} field.
+ *
+ * <p>Without the reset, a subsequent {@code switchToWriteMode()} triggered by a
+ * read-then-write sequence seeks the underlying {@link RandomAccessFile} to the
+ * stale {@code writePosition} (past the just-truncated EOF). The next
+ * {@code file.write(...)} extends the file at that position; the filesystem
+ * zero-fills the range between the truncated length and the stale seek
+ * position, leaving a contiguous zero gap inside an otherwise append-only
+ * binary data file.
+ *
+ * <p>Under TDB2 this surfaces as {@code NodeTableTRDF/Read ->
+ * TProtocolException: Unrecognized type 0} on any {@code nodes-data.obj}
+ * lookup whose NodeId resolves into the zero gap.
+ */
+public class TestBinaryDataFileRandomAccessTruncate {
+
+    private static final String FILE = "target/test-bdfra-truncate-writepos";
+
+    private BinaryDataFile file;
+
+    @BeforeEach public void before() {
+        FileOps.delete(FILE);
+        file = new BinaryDataFileRandomAccess(FILE);
+        file.open();
+    }
+
+    @AfterEach public void after() {
+        file.close();
+        FileOps.delete(FILE);
+    }
+
+    /**
+     * truncate() then write() (no intervening read) — file must contain only
+     * the written bytes up to the new length.
+     */
+    @Test public void truncate_then_write_no_gap() {
+        byte[] block = filled(300 * 1024, (byte) 0xAB);
+        file.write(block);
+        assertEquals(300 * 1024L, file.length());
+
+        file.truncate(100 * 1024);
+        assertEquals(100 * 1024L, file.length());
+
+        byte[] tail = filled(100, (byte) 0xCD);
+        long writtenAt = file.write(tail);
+        assertEquals(100 * 1024L, writtenAt,
+                "write() after truncate() must start at the truncated length, not at a stale writePosition");
+        assertEquals(100 * 1024L + 100, file.length());
+
+        assertNoZeroRun(FILE);
+    }
+
+    /**
+     * truncate() then read() then write() — the read flips the internal
+     * readMode flag, so the subsequent write goes through switchToWriteMode
+     * and seeks to the stored writePosition. Without the fix, that position
+     * is stale and writes past the truncated EOF, zero-filling the gap.
+     */
+    @Test public void truncate_then_read_then_write_no_gap() {
+        byte[] block = filled(300 * 1024, (byte) 0xAB);
+        file.write(block);
+
+        file.truncate(100 * 1024);
+        assertEquals(100 * 1024L, file.length());
+
+        byte[] sample = new byte[16];
+        file.read(0, sample);
+
+        byte[] tail = filled(100, (byte) 0xCD);
+        long writtenAt = file.write(tail);
+        assertEquals(100 * 1024L, writtenAt,
+                "write() after truncate()+read() must start at the truncated length, not at a stale writePosition");
+        assertEquals(100 * 1024L + 100, file.length());
+
+        assertNoZeroRun(FILE);
+    }
+
+    private static byte[] filled(int size, byte v) {
+        byte[] b = new byte[size];
+        Arrays.fill(b, v);
+        return b;
+    }
+
+    /**
+     * Fail the enclosing test if the on-disk file contains a contiguous run
+     * of zero bytes of length >= 16. The source blocks are all 0xAB/0xCD so
+     * any zero byte is evidence of filesystem zero-fill past a stale seek.
+     */
+    private static void assertNoZeroRun(String path) {
+        try (RandomAccessFile raf = new RandomAccessFile(path, "r")) {
+            long total = raf.length();
+            byte[] buf = new byte[(int) total];
+            raf.readFully(buf);
+            int runStart = -1;
+            int runLen = 0;
+            for (int i = 0; i < buf.length; i++) {
+                if (buf[i] == 0) {
+                    if (runStart < 0) { runStart = i; runLen = 1; }
+                    else runLen++;
+                } else {
+                    if (runLen >= 16) {
+                        assertTrue(false,
+                                "Zero-byte run of " + runLen + " bytes starting at offset " + runStart
+                                        + " indicates a filesystem zero-fill (stale writePosition after truncate).");
+                    }
+                    runStart = -1;
+                    runLen = 0;
+                }
+            }
+            if (runLen >= 16) {
+                assertTrue(false,
+                        "Zero-byte run of " + runLen + " bytes at end of file starting at offset " + runStart);
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+}

